### ISSUE TITLE
Suggestions for Export Highlights Plugin

### ISSUE THREAD

**Issue by offset-torque** [koreader/koreader#9273]
Firstly thanks for updating the Exporter plugin. It is very functional now, especially Markdown by Highlight type is a very nice touch.

But there are some usability problems with the plugin.

**1. POPUP**
Export popup dialog makes many mistakes which results with a bad UX. 
It shows:
- A complicated popup
- Without important info emphasized
- For a very short time

When a user perform an export, a popup like this appears:

![currentdialog](https://user-images.githubusercontent.com/84420582/176822674-967977a6-e6fc-4cfc-81ca-91eac556da56.png)

Perception and registering of this popup takes at least one second. 
(Remember that not all of our users are young, sharp programmers so it can be even longer than that)
That leaves 2 seconds to the user to read this blob of text and parse all this information.
Unless they have photographic memory, this is impossible.
Especially parsing a filename which spans 4 lines is very demanding.

A short popup is suitable for binary tasks like `Wifi Connected`.
But if we want to give user more information, we should give them time.

**PROPOSED SOLUTION:** Make this dialog user dismissable (and don't close automatically)

Also information in the popup should be easier to parse.
We are showing that popup to give feedback about:
- Status of the export operation
- Type of export
- Location and filename of the export

So this three piece of information can be presented to the user in a format which is easier to parse at a glance like this:

![proposeddialog](https://user-images.githubusercontent.com/84420582/176822762-8912635e-3d6b-4090-aa18-aec29db710b6.png)

**2. LOCATION**
I don't know the discussions behind the export location, maybe this is some legacy code but current location is unintuitive. 
Unless there is a very good reason, we shouldn't send average user into the KOReader root directory.

Expected location for the export file is to be:
1. Close to the exported book and 
2. Under the user library HOME (where a user interacts with 99% of time).

In our case when a user exports highlights, they would probably expect to find the file in these places:
1. Same directory as the book (not a good idea, cluttering)
3. Same directory as the book but in an **Exported** folder (better than the first but will cause lots of Exported sub-dirs)
4. Under an **Exported** folder directly under the user's HOME folder (library root)

But no one expects the CLIPBOARD directory. 
User didn't even interacted with the clipboard, nothing copied or pasted.

So in order to open this exported file, user has to (in my Kobo):
- Open File Explorer
- Go up
- Go to **.adds**
- Go to **koreader**
- Go to **clipboard**

Is this really necessary ?
We keep all the book settings including highlights with the book (sdr directory).
Why separate the exports ?

**PROPOSED SOLUTION:** Unless there is a good reason, I suggest moving export location to the third option mentioned above.
This will have an additional benefit that when a user syncs their home folder from the device, exports will be copied too.

**3. FILENAME**
Currently filename starts with the full date. 
I don't know what is the reasoning behind this (online sync versioning ?!?) but I think for the majority of the users, date is not the most important information. It is the book name which highlights are exported from.

When a user wants to read some exported highlights and go to their exports folder, they don't think:
"I should open the file that I exported on 25th of December last year"

They would think:
"I should open Alice in Wonderland export"

"Filename first" also makes sense when sorting too, obviously. 
Date can be important for versioning but versioning info is generally appended to the end of the file name.

**PROPOSED SOLUTION:** Unless there is a very good reason, I suggest moving book name to the beginning of the exported file name

Of course solutions I offered are open to discussion but these issues should be addressed one way or another for a good user experience.

**Comment by pazos** [comment]
Hey @offset-torque, thanks for your feedback!

> 1. POPUP

It is even worst. The plugin allows you to export up to 6 different formats and services at the same time. The popup actually shows ok/fail status for each enabled target and the path of each local format if ok


>  Make this dialog user dismissable (and don't close automatically)

Fine. A InfoMessage without a timeout.

> Also information in the popup should be easier to parse.
We are showing that popup to give feedback about:

> Status of the export operation
> Type of export
> Location and filename of the export

Repeated up to 6 or more. Could you make a sketch of how it will look like?

> 2. LOCATION

1. Making it configurable (one single path for all local formats) 
2. Fallback is current default.

> 3. FILENAME
> Currently filename starts with the full date.

It was discussed in the refactor. Reason is sorting. I agree with you with name first but was outnumbered and the reason was sound.

Also filenames tend to be rather long and the path part is always repeated, which makes 1.POPUP even harder to understand.

**Comment by pazos** [comment]
I've also didn't like the result and played a bit with changes

```diff
diff --git a/plugins/exporter.koplugin/base.lua b/plugins/exporter.koplugin/base.lua
index 546b6ff8..86de0f90 100644
--- a/plugins/exporter.koplugin/base.lua
+++ b/plugins/exporter.koplugin/base.lua
@@ -79,6 +79,18 @@ Exports a table of booknotes to local format or remote service
 ]]
 function BaseExporter:export(t) end
 
+--[[--
+Absolute path to the dir where local notes are exported
+
+@treturn string asbolute path or nil
+]]
+
+function BaseExporter:getBasePath()
+    if not self.is_remote then
+        return self.clipping_dir
+    end
+end
+
 --[[--
 File path where the exporter writes its output
 
@@ -87,11 +99,23 @@ File path where the exporter writes its output
 ]]
 function BaseExporter:getFilePath(t)
     if not self.is_remote then
-        local filename = string.format("%s-%s.%s",
+        return string.format("%s-%s.%s",
             self:getTimeStamp(),
             #t == 1 and t[1].title or "all-books",
             self.extension)
-        return self.clipping_dir .. "/" .. getSafeFilename(filename)
+    end
+end
+
+--[[--
+Absolute file path where the exporter writes its output
+
+@param t table of booknotes
+@treturn string absolute path or nil
+]]
+
+function BaseExporter:getAbsolutePath(t)
+    if not self.is_remote then
+        return string.format("%s/%s", self:getBasePath(), getSafeFilename(self:getFilePath(t)))
     end
 end
 
diff --git a/plugins/exporter.koplugin/main.lua b/plugins/exporter.koplugin/main.lua
index 4a0bca05..201f8aca 100644
--- a/plugins/exporter.koplugin/main.lua
+++ b/plugins/exporter.koplugin/main.lua
@@ -26,6 +26,7 @@
 
 local DataStorage = require("datastorage")
 local Device = require("device")
+local Font = require("ui/font")
 local InfoMessage = require("ui/widget/infomessage")
 local InputContainer = require("ui/widget/container/inputcontainer")
 local MyClipping = require("clip")
@@ -177,34 +178,57 @@ function Exporter:exportClippings(clippings)
     local export_callback = function()
         UIManager:nextTick(function()
             local timestamp = os.time()
-            local statuses = {}
+            local summary = {
+                count = 0,
+                ok_count = 0,
+                fail_count = 0,
+                local_count = 0,
+                remote_count = 0,
+                local_msg = {},
+                remote_msg = {},
+            }
+
             for k, v in pairs(self.targets) do
                 if v:isEnabled() then
                     v.timestamp = timestamp
-                    local status = v:export(exportables)
-                    if status then
-                        if v.is_remote then
-                            table.insert(statuses, _(v.name .. ": Exported successfully."))
-                        else
-                            table.insert(statuses, _(v.name .. ": Exported to " ) .. v:getFilePath(exportables))
+                    local ok = v:export(exportables)
+                    summary.count = summary.count + 1
+                    if ok then
+                        summary.ok_count = summary.ok_count + 1
+                    else
+                        summary.fail_count = summary.fail_count + 1
+                    end
+                    if v.is_remote then
+                        summary.remote_count = summary.remote_count + 1
+                    else
+                        summary.local_count = summary.local_count + 1
+                        if not summary.local_path then
+                            summary.local_path = v:getBasePath()
                         end
+                    end
+
+                    if ok and v.is_remote then
+                        summary.remote_msg[v.name] = _("Ok")
+                    elseif ok then
+                        summary.local_msg[v.name] = v:getFilePath(exportables)
+                    elseif v.is_remote then
+                        summary.remote_msg[v.name] = _("Failed")
                     else
-                        table.insert(statuses, _(v.name .. ": Failed to export."))
+                        summary.local_msg[v.name] = _("Failed")
                     end
                     v.timestamp = nil
                 end
             end
-            UIManager:show(InfoMessage:new{
-                text = table.concat(statuses, "\n"),
-                timeout = 3,
+
+            UIManager:show(InfoMessage:new {
+                text = _("Exporting may take several seconds…"),
+                timeout = 1,
             })
-        end)
 
-        UIManager:show(InfoMessage:new {
-            text = _("Exporting may take several seconds…"),
-            timeout = 1,
-        })
+            self:report(summary)
+        end)
     end
+
     if self:requiresNetwork() then
         NetworkMgr:runWhenOnline(export_callback)
     else
@@ -280,4 +304,35 @@ function Exporter:addToMainMenu(menu_items)
     menu_items.exporter = menu
 end
 
+function Exporter:report(t)
+    if t.count == 0 then return end
+
+    local summary = string.format("Exported to %d targets: %d ok, %d failed",
+        t.count, t.ok_count, t.fail_count)
+
+    local details = function(subtable)
+        for name, status in pairs(subtable) do
+            summary = summary .. "\n\n" .. name .. ": " .. status
+        end
+    end
+
+    if t.local_count >= 1 then
+        summary = summary .. "\n\n" .. string.format(_("Local targets (saved in %s)"), t.local_path)
+        --summary = summary .. "\n\n" .. _("Local targets (saved in") .. " " .. t.local_path
+        details(t.local_msg)
+    end
+
+    if t.remote_count >= 1 then
+        summary = summary .. "\n\n" .. _("Remote targets:")
+        details(t.remote_msg)
+    end
+
+    UIManager:show(InfoMessage:new{
+        text = summary,
+        face = Font:getFace("x_smallinfofont"),
+        width = math.floor(Device.screen:getWidth() * 4/5),
+        --show_icon = false,
+    })
+end
+
 return Exporter
diff --git a/plugins/exporter.koplugin/target/html.lua b/plugins/exporter.koplugin/target/html.lua
index 94c71af5..3c5905ad 100644
--- a/plugins/exporter.koplugin/target/html.lua
+++ b/plugins/exporter.koplugin/target/html.lua
@@ -57,7 +57,7 @@ function HtmlExporter:getRenderedContent(t)
 end
 
 function HtmlExporter:export(t)
-    local path = self:getFilePath(t)
+    local path = self:getAbsolutePath(t)
     local file = io.open(path, "w")
     if not file then return false end
     local content = self:getRenderedContent(t)
diff --git a/plugins/exporter.koplugin/target/json.lua b/plugins/exporter.koplugin/target/json.lua
index 4774e4a4..f0fa0424 100644
--- a/plugins/exporter.koplugin/target/json.lua
+++ b/plugins/exporter.koplugin/target/json.lua
@@ -24,7 +24,7 @@ end
 function JsonExporter:export(t)
     local exportable
     local timestamp = self.timestamp or os.time()
-    local path = self:getFilePath(t)
+    local path = self:getAbsolutePath(t)
     if #t == 1 then
         exportable = format(t[1])
         exportable.created_on = timestamp
diff --git a/plugins/exporter.koplugin/target/markdown.lua b/plugins/exporter.koplugin/target/markdown.lua
index 84401da8..49b07eb8 100644
--- a/plugins/exporter.koplugin/target/markdown.lua
+++ b/plugins/exporter.koplugin/target/markdown.lua
@@ -120,7 +120,7 @@ function MarkdownExporter:getMenuTable()
 end
 
 function MarkdownExporter:export(t)
-    local path = self:getFilePath(t)
+    local path = self:getAbsolutePath(t)
     local file = io.open(path, "w")
     if not file then return false end
     for idx, book in ipairs(t) do
diff --git a/plugins/exporter.koplugin/target/text.lua b/plugins/exporter.koplugin/target/text.lua
index f5a5ab43..c23887c1 100644
--- a/plugins/exporter.koplugin/target/text.lua
+++ b/plugins/exporter.koplugin/target/text.lua
@@ -42,7 +42,7 @@ end
 function TextExporter:export(t)
     -- Use wide_space to avoid crengine to treat it specially.
 
-    local path = self:getFilePath(t)
+    local path = self:getAbsolutePath(t)
     local file = io.open(path, "a")
     if not file then return false end
     for __, booknotes in ipairs(t) do
```

But that still doesn't look nice as we're constrained by the limits of the InfoMessage widget. We might need some sort of SummaryWidget with an array of messages.

**Comment by pazos** [comment]
And here's the preview with previous diff applied:

![Sin nombre](https://user-images.githubusercontent.com/975883/176861379-97d3b4c8-7505-48b9-8a4b-ba3ac27c1d0a.png)

Much better but yet not perfect

**Comment by offset-torque** [comment]
Thanks for the quick reply **pazos**

> The plugin allows you to export up to 6 different formats and services at the same time. The popup actually shows ok/fail status for each enabled target and the path of each local format if ok.

Oh I didn't know it can do this. Cool.


> A InfoMessage without a timeout.

Even this is an improvement by itself, new users can easily miss the saving location otherwise.

> It was discussed in the refactor. Reason is sorting. I agree with you with name first but was outnumbered and the reason was sound.

I understand.

> Making it configurable (one single path for all local formats)

Are you planning to add this configuration you mean ? Or did I misunderstood ?

> Repeated up to 6 or more. Could you make a sketch of how it will look like?

Actually your modification looks good. 
Just some bolding and capital letters to attract user's eye to the most important information would be better.
And maybe icons to show the status at a glance.

I don't know InfoMessage widget capabilities so I assumed it can show bold characters and icons
(but I suspect it can't even do bold, right? :)
So here is a bit improved version of your one. 

It doesn't need to be perfect, just a bit clearer to the user is enough.

![proposeddialog2](https://user-images.githubusercontent.com/84420582/176877954-2de345d1-8658-4ad3-86df-cf4077777408.png)

**Comment by pazos** [comment]
> Are you planning to add this configuration you mean ?

yes

> I don't know InfoMessage widget capabilities so I assumed it can show bold characters and icons

nope, neither fancy horizontal splitters like in your sketch.

Hence my suggestion:

> We might need some sort of SummaryWidget with an array of messages.

Where each "message" behaves like an InfoMessage (with its own icon) and they're stacked vertically.

Anyhow, I'm not an UX guy and I don't have too much knowledge* of our UI toolbox  (*I mean, I barely know InfoMessage et al), so pinging @poire-z, who knows way better, if my answer is wrong.

I'm interested more in features and display them in a way it makes *some* sense. Still not convinced with both my and your previews. There're some chunks of code repeated.

Also this plugin now it's a complex beast: it can be expanded to many more targets without touching much of the sharing code, so I would like to figure out a way that works now and works with *n* targets.

To sum up:

1. Two different kind of targets: local files and remote services.
2. Locales files are placed on a path and have a filename.
3. The path and filename is equal for all the local files for a given export, only the extension changes between formats.

Some idiomatic one-liner should contain all the info that's relevant:

`Local: saved in ./clipboard/YYYY-DD-MM-HH-MM-SS-my-lenghty-filename-with-a-bunch-of-words.{md,html,txt,json,sqlite,yml,wewewe}`

And probably per format info is only relevant in case of an error

I don't really know but it is on my pile of stuff to do :)

**Comment by poire-z** [comment]
InfoMessage should auto-reduce the font size so it can show it all on the screen. (ie. the long text from tap on (i) icon in Book map).
Otherwise, there is TextViewer that will not reduce the font size, but use a vertical scrollbar.
Both can't do bold - but you could use some Unicode symbol to show success or failure.

Othewise, you could build a more complex widget with stacked widgets (like Coverbrowser list view or the new vocabulary builder) - but I don't think it's worth the time and pain building something that complex for just a status report.

**Comment by offset-torque** [comment]
> I don't think it's worth the time and pain building something that complex for just a status report.

I agree with poire-z about this.

Considering the limitations of the widget (we only have vertical separation for positioning and none of the styling) and your expandability requirement, I think something like this is the best we can do. It is not very glanceable nor pretty but at least it stays on the screen now and users can analyze it until he gets familiar with the structure:

![proposeddialog3](https://user-images.githubusercontent.com/84420582/176909872-9ddba119-4b29-4f22-b9a5-098ea2138313.png)

**Comment by offset-torque** [comment]
I used the export plugin on my device today (I generally use KoHighlights to export them) and saw that short duration of export popup persists.

**Reminder from the first post:**

> Perception and registering of this popup takes at least one second.
> That leaves 2 seconds to the user to read this blob of text and parse all this information.
> Unless they have photographic memory, this is impossible.
>  **PROPOSED SOLUTION:** Keep this dialog open until user dismisses it

I had to export two times to be able to peek at the export location.
This is still a usability issue and probably not that hard to fix.

**Comment by hius07** [comment]
Maybe change the target folder name from "clipboard" to "export"?

**Comment by offset-torque** [comment]
> **hius07** Maybe change the target folder name from "clipboard" to "export"?

If we are at it, also location of this directory is wrong.
I don't know the historical reasons but we are sending users to the koreader installation tree to open their exported files. This is not ok.

When a user exports a bookmark file, logically expected location is generally the "same directory as the book file" or the "main directory of the library under a name like **Exports**" to avoid clutter.

**If I export bookmarks and later want to have a look at that file this is a logical workflow:**
1. Open **File Browser**. My **Library** folder is showing now.
2. Tap **Exports** directory
3. Open exported file


**Current steps:**
1. Open **File Browser**
2. Open the **Menu**
3. Tap **Settings**
4. Tap "**Show hidden files**"
5. Close menu
6. Go up one directory
7. Tap **.adds** directory
8. Tap **.koreader** directory
9. Tap **.clipboard** directory
10. Open exported file

Of course I assumed that user knows how to traverse through koreader installation, which is a very generous assumption. And I also assumed that they memorized this location from the 2 second popup which is even more generous.

This is bad UX. There is nothing intuitive in this steps. I am curious what is the benefit of using that directory to justify this inconvenience. Why my exports can't be under my "Library home" which is the expected and logical place for them.

**Comment by hius07** [comment]
What is the location suggested for readonly book storage (sd card under Android)?
We can add a user setting to choose the exporting folder (like for screenshots).

**Comment by pazos** [comment]
> I don't know the historical reasons but we are sending users to the koreader installation tree to open their exported files. This is not ok.

It used to be a plugin that exported highlights to evernote cloud, made by clever people. So not important back then.
Other people added other targets on top of the same development.

The refactor we made to make the code maintenable is intended to make way easier to address this kind of issues. We didn't change the "inner" behaviour of the plugin, including hardcoded paths, just the structure of the code.

Now, the suggestion @offset-torque made is valid (and was requested a few times in mobileread, too). Please make it as part of the BaseExporter, so it can be inherited by all **local** targets without duplicating code. Let's assume the same path works for all kind of exports for the same user.

A file chooser does the job. IMHO the old path should be preserved as default until the user chooses a different one, so we don't mess with current users of the plugin that are doing ok with the overcomplicated workflow we have now :)

**Comment by offset-torque** [comment]
> IMHO the old path should be preserved as default until the user chooses a different one, so we don't mess with current users of the plugin that are doing ok with the overcomplicated workflow we have now :)

That is exactly the problem because future users will still be exposed to this problem. If we are keeping the old directory as default, there is no sense of implementing a chooser because new users will not even be aware of the fact that they can change this directory (unless you have a way to tell them about this via some dialog or smt like that). 

As a programmer you are aware of the **sane defaults** principle. "People in the past decided like this so we can't change this code" can't be an excuse for bad UX. Is there an API or automated workflow that you are afraid of breaking with this change ?

**Here is a solution that will protect your current users (who enjoys going around directories on slow devices apparently :)**
- KOReader checks if **Clipboard** directory has exports in it
- If **yes** don't change the default directory assuming this is an old user
- If there are **no** exports already there, change the default to **Exports** assuming this is a new user

**All new users will be automatically migrated. All old users are preserved.**

**Comment by pazos** [comment]
Whatever, dude.

**Comment by offset-torque** [comment]
> Whatever, dude.

Since this is the least professional reply I got from any KOReader developer during my time here, I have no idea what it means, Martin dude. If you can express yourself in complete sentences, people finding this issue with similar complaints would be enlightened too about where this is going. If you have no intention to fix these things, say it clearly so maybe someone else picks it up.

**Comment by pazos** [comment]
I'm sorry.

It means *Do whatever you want. I don't care. I'm not really interested in arguing with you about this nitpick*

> If you have no intention to fix these things, say it clearly so maybe someone else picks it up.

I don't plan to work on it. I don't owe you or anybody an explaination. I'm not assigned to this ticket. What the hell?

If somebody submits a patch that addresses some of the topics discussed above I would gladly review it, if my free time allows it.

**Comment by offset-torque** [comment]
> I don't plan to work on it. I don't owe you or anybody an explaination. I'm not assigned to this ticket. What the hell?

Then why the hell are you here polluting the issue ? You are the one who replied to this issue and half of the messages here belongs to you. Is someone forcing you to write here ?

For one tenth of the effort you spent writing here and complaining, you could have increased this timeout value to a reasonable one and solved the problem:

> UIManager:show(InfoMessage:new{
> text = table.concat(statuses, "\n"),
> **timeout = 3,**

But Martin already knows where his exports are so why would he care ? 
If the duration is too short to read where the hell are the exports gone, users can export again and again until they catch the directory name or go to mobileread forums and ask someone there. (No, of course we can't give them enough time to read the dialog properly, don't be ridiculous.)

I am glad you are the only one with this general attitude amongst the KOReader devs, dude. 
Don't waste your precious time here anymore, it's all nitpicks here.

**Comment by pazos** [comment]
Wow. I see I really hurted your feelings with that "dude". I'm going to repeat my "I'm sorry" with a bold typeface **My apologies, It wasn't my intention**. 

And to be clear, the whatever/nitpick was refering to your previous comment https://github.com/koreader/koreader/issues/9273#issuecomment-1420812077, where you suggested some workflow that doesn't change, improve or make easier any of your remarks in https://github.com/koreader/koreader/issues/9273#issuecomment-1420488436.

While I appreciated your initial feedback and tried to provide some info and a PoC to address some of the issues you've remarked, I'm not always the best version of myself and that "whatever, dude" was my way to tell you I don't care enough to argue against your suggestion.

Next time I will be silent until I have something important to say :)

Now, please, if you need to address some other things with me I'm on gitter and mobileread. We can chat there if you wish. Not here. I repeat, not here, please.

**Comment by offset-torque** [comment]
@hius07 
Thanks for implementing my suggestion. I was thinking of rewriting the Highlighting and Export sections. I will add this feature to the guide after testing it myself.

---

@pazos 
For some reason I didn't receive a notification for your last reply so I am seeing it now. I understand that you are not under any obligation to implement the suggestions but you can disagree and refuse while still being respectful. 

In addition to being a user, I wrote the user guide and I am updating it for 2 years so I am a contributor to this project. Also there are now many features and fixes that is implemented from my suggestions. I don't think you talk to other contributors in this tone. Maybe you were having a bad day, I don't know. I don't have a personal problem with you and I hope you don't have one with me. 

Thanks for the sincere apology, I suggest a reset to avoid future interactions like this. This is really unnecessary since we are here for the same purpose. As long as your attitude is respectful and reasonable, I assure you I will respond in the same manner.

**Comment by offset-torque** [comment]
I am cleaning up the old open issues. Since **hius07** implemented the **"choosing export folder"** functionality and I also added this to the user guide, users now can reach this information from within the KOReader or via online help. So main complaint of this issue (impossibility to see the export location in a short duration, crowded popup) is partially solved and I am closing it.

But I want to remind that, export popup is still very bad and this version -> https://github.com/koreader/koreader/issues/9273#issuecomment-1172108094 you made above is still much better than the current one **pazos**. So I suggest at least implementing that one, since you already wrote the code.
